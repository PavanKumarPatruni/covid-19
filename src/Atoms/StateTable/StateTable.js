import React, { useState } from 'react';
import PropTypes from 'prop-types';

import ArrowUp from '../ArrowUp';
import ShowToggle from '../ShowToggle';

import { Colors } from '../../utils/contants';

import './StateTable.scss';

const propTypes = {
  stateList: PropTypes.array,
  districts: PropTypes.object,
};

const StateTable = ({ stateList, districts }) => {
  const [selected, setSelected] = useState('');

  const getData = (item, key) => {
    return item && item[key];
  };

  const getTotal = () => {
    return stateList[0];
  };

  const getUpArrow = isGreen => {
    return <ArrowUp color={isGreen ? Colors.GREEN : Colors.RED} size="12" />;
  };

  const getDeltaComponent = (item, key, isGreen) => {
    const value = getData(item, key);

    return value !== '0' && value > 0 ? (
      <div className="flex flex-center">
        {getUpArrow(isGreen)}
        <p className={`tag mr1 ${isGreen ? 'tag-green' : ''}`}>{value}</p>
      </div>
    ) : null;
  };

  const onStateClick = stateCode => {
    if (stateCode === 'TT' || stateCode === selected) {
      setSelected('');
      return;
    }
    setSelected(stateCode);
  };

  const getDistrictData = (key, district) => {
    return (
      <tr key={key} className="district-row">
        <td className="text-left">{key}</td>
        <td>
          <div className="flex flex-end flex-justify-end sm-flex-wrap sm-flex-justify-center">
            {getDeltaComponent(district.delta, 'confirmed')}
            <p>{getData(district, 'confirmed')}</p>
          </div>
        </td>
        <td>{getData(district, 'active')}</td>
        <td>
          <div className="flex flex-end flex-justify-end sm-flex-wrap sm-flex-justify-center">
            {getDeltaComponent(district.delta, 'recovered', true)}
            <p>{getData(district, 'recovered')}</p>
          </div>
        </td>
        <td>
          <div className="flex flex-end flex-justify-end sm-flex-wrap sm-flex-justify-center">
            {getDeltaComponent(district.delta, 'deceased')}
            <p>{getData(district, 'deceased')}</p>
          </div>
        </td>
      </tr>
    );
  };

  const getDistricts = (key, item) => {
    if (!item) {
      return null;
    }

    const districtListObjects = item.districtData;
    const districtList = Object.keys(districtListObjects);
    if (!districtList) {
      return null;
    }

    return districtList.map(district =>
      selected === item.statecode
        ? getDistrictData(district, districtListObjects[district])
        : null
    );
  };

  const getArrowHandler = statecode => {
    return (
      <ShowToggle
        color={Colors.VOILET}
        size={16}
        show={selected === statecode}
      />
    );
  };

  const getRowData = item => {
    const name = getData(item, 'state');
    const confirmed = getData(item, 'confirmed');

    const stateArray = [];
    stateArray.push(
      <tr
        key={item.statecode}
        onClick={() => onStateClick(item.statecode)}
        id={`state-row-${item.statecode}`}
        className={`${selected === item.statecode ? ' state-row' : ''}`}
      >
        <td className="text-left flex flex-center flex-space-between">
          <p className="full-width">{name}</p>
          {item.statecode !== 'TT' && confirmed > 0
            ? getArrowHandler(item.statecode)
            : null}
        </td>
        <td className="text-center">
          <div className="flex flex-end flex-justify-end sm-flex-wrap sm-flex-justify-center">
            {getDeltaComponent(item, 'deltaconfirmed')}
            <p>{confirmed}</p>
          </div>
        </td>
        <td className="text-center">{getData(item, 'active')}</td>
        <td className="text-center">
          <div className="flex flex-end flex-justify-end sm-flex-wrap sm-flex-justify-center">
            {getDeltaComponent(item, 'deltarecovered', true)}
            <p>{getData(item, 'recovered')}</p>
          </div>
        </td>
        <td className="text-center">
          <div className="flex flex-end flex-justify-end sm-flex-wrap sm-flex-justify-center">
            {getDeltaComponent(item, 'deltadeaths')}
            <p>{getData(item, 'deaths')}</p>
          </div>
        </td>
      </tr>
    );

    const districtArray =
      item.statecode === selected ? getDistricts(name, districts[name]) : [];
    return [...stateArray, ...(districtArray || [])];
  };

  const getRows = () => {
    return stateList
      .filter(item => item.statecode !== 'TT')
      .map(item => getRowData(item));
  };

  const getTotalRow = () => {
    return getRowData(getTotal());
  };

  return stateList && stateList.length > 0 ? (
    <table className="sm-p0 comp-mb">
      <thead>
        <tr>
          <th className="text-left">State</th>
          <th className="text-center red">Confirmed</th>
          <th className="text-center blue">Active</th>
          <th className="text-center green">Recovered</th>
          <th className="text-center voilet">Deceased</th>
        </tr>
      </thead>
      <tbody>
        {getRows()}
        {getTotalRow()}
      </tbody>
    </table>
  ) : null;
};

StateTable.defaultProps = {
  stateList: [],
  districts: {},
};

StateTable.propTypes = propTypes;

export default StateTable;
