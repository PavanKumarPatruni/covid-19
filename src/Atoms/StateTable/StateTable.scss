@import '../../styles/index.scss';

table {
  padding: 16px;
}

th,
td {
  border-radius: 8px;
  color: $text-color;
  margin: 0 4px;
  padding: 8px;
  text-align: center;
}

th {
  @include table-heading($text-color);

  text-transform: uppercase;
}

th:first-child {
  text-align: left;
}

td {
  @include table-row($text-color);
}

td:first-child {
  @include bold();

  text-align: left;
  width: 200px;
}

tr {
  background: $table-background;
}

tr:hover {
  background: $border;
}

.tag {
  @include table-heading($red-color);
}

.tag-green {
  @include table-heading($green-color);
}

.up-arrow-icon {
  fill: $red-color;
  height: 12px;
  stroke: $red-color;
  width: 12px;
}

.green-icon {
  fill: $green-color;
  stroke: $green-color;
}
