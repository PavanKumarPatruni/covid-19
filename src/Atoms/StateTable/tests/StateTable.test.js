import React from 'react';
import { shallow, mount } from 'enzyme';

import StateTable from '../StateTable';

describe('StateTable component', () => {
  const props = {
    stateList: [
      {
        active: '44351',
        confirmed: '67700',
        deaths: '2215',
        deltaconfirmed: '524',
        deltadeaths: '2',
        deltarecovered: '160',
        lastupdatedtime: '11/05/2020 15:41:22',
        recovered: '21130',
        state: 'Total',
        statecode: 'TT',
        statenotes: '',
      },
      {
        active: '44351',
        confirmed: '67700',
        deaths: '2215',
        deltaconfirmed: '524',
        deltadeaths: '2',
        deltarecovered: '160',
        lastupdatedtime: '11/05/2020 15:41:22',
        recovered: '21130',
        state: 'Total',
        statecode: 'AP',
        statenotes: '',
      },
      {
        active: '44351',
        confirmed: '67700',
        deaths: '2215',
        deltaconfirmed: '524',
        deltadeaths: '0',
        deltarecovered: '160',
        lastupdatedtime: '11/05/2020 15:41:22',
        recovered: '21130',
        state: 'Total',
        statecode: 'TN',
        statenotes: '',
      },
    ],
  };

  test('match snapshot correctly', () => {
    const wrapper = shallow(<StateTable {...props} />);
    expect(wrapper).toMatchSnapshot();
  });

  test('match snapshot correctly', () => {
    const wrapper = mount(<StateTable {...props} />);
    expect(wrapper).toMatchSnapshot();
  });

  test('match snapshot correctly- with empty array', () => {
    props.stateList = [];
    const wrapper = shallow(<StateTable {...props} />);
    expect(wrapper).toMatchSnapshot();
  });

  test('match snapshot correctly - with delta value "0"', () => {
    props.stateList = [];
    const wrapper = mount(<StateTable {...props} />);
    expect(wrapper).toMatchSnapshot();
  });
});
