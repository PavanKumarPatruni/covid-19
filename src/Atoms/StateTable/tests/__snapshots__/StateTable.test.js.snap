// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StateTable component match snapshot correctly - with delta value "0" 1`] = `
<StateTable
  stateList={Array []}
/>
`;

exports[`StateTable component match snapshot correctly 1`] = `
<table
  className="fadeInUp sm-p0"
>
  <thead>
    <tr>
      <th>
        State
      </th>
      <th>
        Confirmed
      </th>
      <th>
        Active
      </th>
      <th>
        Recovered
      </th>
      <th>
        Deceased
      </th>
    </tr>
  </thead>
  <tbody>
    <tr
      key="AP"
    >
      <td>
        Total
      </td>
      <td>
        <div
          className="flex flex-end flex-justify-end"
        >
          <div
            className="flex flex-end"
          >
            <ArrowUp
              color="#ff073a"
              size="14"
            />
            <p
              className="tag mr1 "
            >
              524
            </p>
          </div>
          <p>
            67700
          </p>
        </div>
      </td>
      <td>
        44351
      </td>
      <td>
        <div
          className="flex flex-end flex-justify-end"
        >
          <div
            className="flex flex-end"
          >
            <ArrowUp
              color="#28a745"
              size="14"
            />
            <p
              className="tag mr1 tag-green"
            >
              160
            </p>
          </div>
          <p>
            21130
          </p>
        </div>
      </td>
      <td>
        <div
          className="flex flex-end flex-justify-end"
        >
          <div
            className="flex flex-end"
          >
            <ArrowUp
              color="#ff073a"
              size="14"
            />
            <p
              className="tag mr1 "
            >
              2
            </p>
          </div>
          <p>
            2215
          </p>
        </div>
      </td>
    </tr>
    <tr
      key="TN"
    >
      <td>
        Total
      </td>
      <td>
        <div
          className="flex flex-end flex-justify-end"
        >
          <div
            className="flex flex-end"
          >
            <ArrowUp
              color="#ff073a"
              size="14"
            />
            <p
              className="tag mr1 "
            >
              524
            </p>
          </div>
          <p>
            67700
          </p>
        </div>
      </td>
      <td>
        44351
      </td>
      <td>
        <div
          className="flex flex-end flex-justify-end"
        >
          <div
            className="flex flex-end"
          >
            <ArrowUp
              color="#28a745"
              size="14"
            />
            <p
              className="tag mr1 tag-green"
            >
              160
            </p>
          </div>
          <p>
            21130
          </p>
        </div>
      </td>
      <td>
        <div
          className="flex flex-end flex-justify-end"
        >
          <p>
            2215
          </p>
        </div>
      </td>
    </tr>
    <tr
      key="TT"
    >
      <td>
        Total
      </td>
      <td>
        <div
          className="flex flex-end flex-justify-end"
        >
          <div
            className="flex flex-end"
          >
            <ArrowUp
              color="#ff073a"
              size="14"
            />
            <p
              className="tag mr1 "
            >
              524
            </p>
          </div>
          <p>
            67700
          </p>
        </div>
      </td>
      <td>
        44351
      </td>
      <td>
        <div
          className="flex flex-end flex-justify-end"
        >
          <div
            className="flex flex-end"
          >
            <ArrowUp
              color="#28a745"
              size="14"
            />
            <p
              className="tag mr1 tag-green"
            >
              160
            </p>
          </div>
          <p>
            21130
          </p>
        </div>
      </td>
      <td>
        <div
          className="flex flex-end flex-justify-end"
        >
          <div
            className="flex flex-end"
          >
            <ArrowUp
              color="#ff073a"
              size="14"
            />
            <p
              className="tag mr1 "
            >
              2
            </p>
          </div>
          <p>
            2215
          </p>
        </div>
      </td>
    </tr>
  </tbody>
</table>
`;

exports[`StateTable component match snapshot correctly 2`] = `
<StateTable
  stateList={
    Array [
      Object {
        "active": "44351",
        "confirmed": "67700",
        "deaths": "2215",
        "deltaconfirmed": "524",
        "deltadeaths": "2",
        "deltarecovered": "160",
        "lastupdatedtime": "11/05/2020 15:41:22",
        "recovered": "21130",
        "state": "Total",
        "statecode": "TT",
        "statenotes": "",
      },
      Object {
        "active": "44351",
        "confirmed": "67700",
        "deaths": "2215",
        "deltaconfirmed": "524",
        "deltadeaths": "2",
        "deltarecovered": "160",
        "lastupdatedtime": "11/05/2020 15:41:22",
        "recovered": "21130",
        "state": "Total",
        "statecode": "AP",
        "statenotes": "",
      },
      Object {
        "active": "44351",
        "confirmed": "67700",
        "deaths": "2215",
        "deltaconfirmed": "524",
        "deltadeaths": "0",
        "deltarecovered": "160",
        "lastupdatedtime": "11/05/2020 15:41:22",
        "recovered": "21130",
        "state": "Total",
        "statecode": "TN",
        "statenotes": "",
      },
    ]
  }
>
  <table
    className="fadeInUp sm-p0"
  >
    <thead>
      <tr>
        <th>
          State
        </th>
        <th>
          Confirmed
        </th>
        <th>
          Active
        </th>
        <th>
          Recovered
        </th>
        <th>
          Deceased
        </th>
      </tr>
    </thead>
    <tbody>
      <tr
        key="AP"
      >
        <td>
          Total
        </td>
        <td>
          <div
            className="flex flex-end flex-justify-end"
          >
            <div
              className="flex flex-end"
            >
              <ArrowUp
                color="#ff073a"
                size="14"
              >
                <svg
                  fill="#ff073a"
                  height="14"
                  stroke="#ff073a"
                  strokeLinecap="round"
                  strokeLinejoin="round"
                  strokeWidth="2"
                  viewBox="0 0 24 24"
                  width="14"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <line
                    x1="12"
                    x2="12"
                    y1="19"
                    y2="5"
                  />
                  <polyline
                    points="5 12 12 5 19 12"
                  />
                </svg>
              </ArrowUp>
              <p
                className="tag mr1 "
              >
                524
              </p>
            </div>
            <p>
              67700
            </p>
          </div>
        </td>
        <td>
          44351
        </td>
        <td>
          <div
            className="flex flex-end flex-justify-end"
          >
            <div
              className="flex flex-end"
            >
              <ArrowUp
                color="#28a745"
                size="14"
              >
                <svg
                  fill="#28a745"
                  height="14"
                  stroke="#28a745"
                  strokeLinecap="round"
                  strokeLinejoin="round"
                  strokeWidth="2"
                  viewBox="0 0 24 24"
                  width="14"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <line
                    x1="12"
                    x2="12"
                    y1="19"
                    y2="5"
                  />
                  <polyline
                    points="5 12 12 5 19 12"
                  />
                </svg>
              </ArrowUp>
              <p
                className="tag mr1 tag-green"
              >
                160
              </p>
            </div>
            <p>
              21130
            </p>
          </div>
        </td>
        <td>
          <div
            className="flex flex-end flex-justify-end"
          >
            <div
              className="flex flex-end"
            >
              <ArrowUp
                color="#ff073a"
                size="14"
              >
                <svg
                  fill="#ff073a"
                  height="14"
                  stroke="#ff073a"
                  strokeLinecap="round"
                  strokeLinejoin="round"
                  strokeWidth="2"
                  viewBox="0 0 24 24"
                  width="14"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <line
                    x1="12"
                    x2="12"
                    y1="19"
                    y2="5"
                  />
                  <polyline
                    points="5 12 12 5 19 12"
                  />
                </svg>
              </ArrowUp>
              <p
                className="tag mr1 "
              >
                2
              </p>
            </div>
            <p>
              2215
            </p>
          </div>
        </td>
      </tr>
      <tr
        key="TN"
      >
        <td>
          Total
        </td>
        <td>
          <div
            className="flex flex-end flex-justify-end"
          >
            <div
              className="flex flex-end"
            >
              <ArrowUp
                color="#ff073a"
                size="14"
              >
                <svg
                  fill="#ff073a"
                  height="14"
                  stroke="#ff073a"
                  strokeLinecap="round"
                  strokeLinejoin="round"
                  strokeWidth="2"
                  viewBox="0 0 24 24"
                  width="14"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <line
                    x1="12"
                    x2="12"
                    y1="19"
                    y2="5"
                  />
                  <polyline
                    points="5 12 12 5 19 12"
                  />
                </svg>
              </ArrowUp>
              <p
                className="tag mr1 "
              >
                524
              </p>
            </div>
            <p>
              67700
            </p>
          </div>
        </td>
        <td>
          44351
        </td>
        <td>
          <div
            className="flex flex-end flex-justify-end"
          >
            <div
              className="flex flex-end"
            >
              <ArrowUp
                color="#28a745"
                size="14"
              >
                <svg
                  fill="#28a745"
                  height="14"
                  stroke="#28a745"
                  strokeLinecap="round"
                  strokeLinejoin="round"
                  strokeWidth="2"
                  viewBox="0 0 24 24"
                  width="14"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <line
                    x1="12"
                    x2="12"
                    y1="19"
                    y2="5"
                  />
                  <polyline
                    points="5 12 12 5 19 12"
                  />
                </svg>
              </ArrowUp>
              <p
                className="tag mr1 tag-green"
              >
                160
              </p>
            </div>
            <p>
              21130
            </p>
          </div>
        </td>
        <td>
          <div
            className="flex flex-end flex-justify-end"
          >
            <p>
              2215
            </p>
          </div>
        </td>
      </tr>
      <tr
        key="TT"
      >
        <td>
          Total
        </td>
        <td>
          <div
            className="flex flex-end flex-justify-end"
          >
            <div
              className="flex flex-end"
            >
              <ArrowUp
                color="#ff073a"
                size="14"
              >
                <svg
                  fill="#ff073a"
                  height="14"
                  stroke="#ff073a"
                  strokeLinecap="round"
                  strokeLinejoin="round"
                  strokeWidth="2"
                  viewBox="0 0 24 24"
                  width="14"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <line
                    x1="12"
                    x2="12"
                    y1="19"
                    y2="5"
                  />
                  <polyline
                    points="5 12 12 5 19 12"
                  />
                </svg>
              </ArrowUp>
              <p
                className="tag mr1 "
              >
                524
              </p>
            </div>
            <p>
              67700
            </p>
          </div>
        </td>
        <td>
          44351
        </td>
        <td>
          <div
            className="flex flex-end flex-justify-end"
          >
            <div
              className="flex flex-end"
            >
              <ArrowUp
                color="#28a745"
                size="14"
              >
                <svg
                  fill="#28a745"
                  height="14"
                  stroke="#28a745"
                  strokeLinecap="round"
                  strokeLinejoin="round"
                  strokeWidth="2"
                  viewBox="0 0 24 24"
                  width="14"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <line
                    x1="12"
                    x2="12"
                    y1="19"
                    y2="5"
                  />
                  <polyline
                    points="5 12 12 5 19 12"
                  />
                </svg>
              </ArrowUp>
              <p
                className="tag mr1 tag-green"
              >
                160
              </p>
            </div>
            <p>
              21130
            </p>
          </div>
        </td>
        <td>
          <div
            className="flex flex-end flex-justify-end"
          >
            <div
              className="flex flex-end"
            >
              <ArrowUp
                color="#ff073a"
                size="14"
              >
                <svg
                  fill="#ff073a"
                  height="14"
                  stroke="#ff073a"
                  strokeLinecap="round"
                  strokeLinejoin="round"
                  strokeWidth="2"
                  viewBox="0 0 24 24"
                  width="14"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <line
                    x1="12"
                    x2="12"
                    y1="19"
                    y2="5"
                  />
                  <polyline
                    points="5 12 12 5 19 12"
                  />
                </svg>
              </ArrowUp>
              <p
                className="tag mr1 "
              >
                2
              </p>
            </div>
            <p>
              2215
            </p>
          </div>
        </td>
      </tr>
    </tbody>
  </table>
</StateTable>
`;

exports[`StateTable component match snapshot correctly- with empty array 1`] = `""`;
